### ISSUE TITLE
cvmfs repository stopped getting updates after cvmfs client 2.11.0 update

### ISSUE THREAD

**Issue by smuzaffar** [cvmfs/cvmfs#3384]
He,
  On few (4 out of 50) of our OpenStack based build nodes at CERN, we noticed that after the update to `cvmfs-2.11.0` some of repositories stop getting updates. System logs show errors messages like [a]. All our build nodes are identical so no idea why this happened on only few nodes. Running `cvmfs_config killall; cvmfs_config reload` fixed the issue though

[a]
```
Sep  4 17:40:15 cmsbuild126.cern.ch cvmfs2[15200]: (singularity.opensciencegrid.org) switching host from http://klei.nikhef.nl:8000/cvmfs/singularity.opensciencegrid.org to http://cvmfs-egi.gridpp.rl.ac.uk:8000/cvmfs/singularity.opensciencegrid.org (reset host)
Sep  4 17:40:15 buildnode cvmfs2[6596]: (unpacked.cern.ch) switching host from http://cernvmfs.gridpp.rl.ac.uk/cvmfs/unpacked.cern.ch to http://cvmfs-stratum-one.cern.ch/cvmfs/unpacked.cern.ch (reset host)
Sep  4 17:40:15 buildnode cvmfs2[6596]: (unpacked.cern.ch) whitelist verification failed (6): invalid whitelist signature
Sep  4 17:40:15 buildnode cvmfs2[6596]: (unpacked.cern.ch) failed to fetch manifest (8 - bad whitelist), trying another stratum 1
Sep  4 17:40:15 buildnode cvmfs2[6596]: (unpacked.cern.ch) switching host from http://cvmfs-stratum-one.cern.ch/cvmfs/unpacked.cern.ch to http://cernvmfs.gridpp.rl.ac.uk/cvmfs/unpacked.cern.ch (manually triggered)
Sep  4 17:40:18 buildnode cvmfs2[6325]: (cms-ib.cern.ch) whitelist verification failed (6): invalid whitelist signature
Sep  4 17:40:18 buildnode cvmfs2[6325]: (cms-ib.cern.ch) failed to fetch manifest (8 - bad whitelist), trying another stratum 1
Sep  4 17:40:18 buildnode cvmfs2[6325]: (cms-ib.cern.ch) switching host from http://cvmfs-stratum-one.cern.ch/cvmfs/cms-ib.cern.ch to http://cvmfs.fnal.gov/cvmfs/cms-ib.cern.ch (manually triggered)
Sep  4 17:40:19 buildnode cvmfs2[6325]: (cms-ib.cern.ch) whitelist verification failed (6): invalid whitelist signature
Sep  4 17:40:19 buildnode cvmfs2[6325]: (cms-ib.cern.ch) warning, could not apply updated catalog revision, entering offline mode
```

**Comment by jblomer** [comment]
@HereThereBeDragons Could this be related to the new sinks in the download manager? Getting the whitelist is one of the few places where we download into a memory sink. Perhaps there is an issue with host failover and the memory sink, e.g. perhaps the memory is not reset properly.

Edit: nothing obvious on first sight. Still worth taking a closer look, I think.

**Comment by jblomer** [comment]
@smuzaffar According to the logs, cvmfs enters "offline mode" (which it shouldn't). After a few minutes, cvmfs should retry if there is again network connection.  Can you tell whether these nodes apply the update eventually, though delayed?

**Comment by HereThereBeDragons** [comment]
i double-check and compared against 2.10.1 for files
```
whitelist.cc
manifest_fetch.cc
```
for proper memsink release/ownership handover

the error seems to come from `manifest_fetch.cc` `DoVerify()` -> `whitelist.LoadUrl(base_url);`
 -> `VerifyWhitelist();`  --> `retval_b = signature_manager_->VerifyLetter(plain_buf_, plain_size_, true);`
which then returns `kFailInvalidCertificate`


one thing that change is in `manifest_fetch.cc` 

```diff
Failures Fetch(const std::string &base_url, const std::string &repository_name,
               const uint64_t minimum_timestamp, const shash::Any *base_catalog,
               signature::SignatureManager *signature_manager,
               download::DownloadManager *download_manager,
               ManifestEnsemble *ensemble) {
  Failures result =
      DoFetch(base_url, repository_name, minimum_timestamp, base_catalog,
              signature_manager, download_manager, ensemble);
  if ((result != kFailOk) &&
      (result != kFailLoad) &&
+      (result != kFailInvalidCertificate) &&
      (download_manager->num_hosts() > 1))
  {
    LogCvmfs(kLogCache, kLogDebug | kLogSyslogWarn,
             "failed to fetch manifest (%d - %s), trying another stratum 1",
             result, Code2Ascii(result));
    download_manager->SwitchHost();
    result = DoFetch(base_url, repository_name, minimum_timestamp, base_catalog,
                     signature_manager, download_manager, ensemble);
  }
  return result;
}
```
so are you sure that this was a 2.11 client? if i understand the `if` correctly you should not go into the `if` and retry with this return type

**Comment by smuzaffar** [comment]
yes it was 2.11 client. cvmfs 2.11 was installed on this node on `Sep  4 17:06:48` and we start getting this error at `Sep  4 17:40:15`. Node did not get any update till `Sep  6 08:42:32` ( this is when I ran `cvmfs_config killall`).

[a]
```
Sep  4 17:06:48 cmsbuild126.cern.ch dnf-distro-sync[3401634]: Running transaction
Sep  4 17:06:49 buildnode dnf-distro-sync[3401634]:  Preparing        :                                                        1/1
Sep  4 17:06:50 buildnode dnf-distro-sync[3401634]:  Running scriptlet: puppet-agent-extra-gems-7.2.1-1.al8.cern.x86_64        1/1
Sep  4 17:06:51 buildnode dnf-distro-sync[3401634]:  Upgrading        : puppet-agent-extra-gems-7.2.1-1.al8.cern.x86_64        1/7
Sep  4 17:06:51 buildnode dnf-distro-sync[3401634]:  Running scriptlet: puppet-agent-extra-gems-7.2.1-1.al8.cern.x86_64        1/7
Sep  4 17:06:58 buildnode dnf-distro-sync[3401634]:  Running scriptlet: puppet-agent-7.24.0.8.gb172499a8-2.al8.cern.x86_64     2/7
Sep  4 17:06:58 buildnode dnf-distro-sync[3401634]:  Upgrading        : puppet-agent-7.24.0.8.gb172499a8-2.al8.cern.x86_64     2/7
Sep  4 17:07:00 buildnode dnf-distro-sync[3401634]:  Running scriptlet: puppet-agent-7.24.0.8.gb172499a8-2.al8.cern.x86_64     2/7
Sep  4 17:07:01 buildnode systemd[1]: Starting Send request to specified server, indicating server is alive and well...
Sep  4 17:07:01 buildnode dnf-distro-sync[3401634]:  Installing       : cvmfs-libs-2.11.0-1.el8.x86_64                         3/7
Sep  4 17:07:01 buildnode dnf-distro-sync[3401634]:  Running scriptlet: cvmfs-libs-2.11.0-1.el8.x86_64                         3/7
Sep  4 17:07:01 buildnode systemd[1]: nc_heartbeat.service: Succeeded.
Sep  4 17:07:01 buildnode systemd[1]: Started Send request to specified server, indicating server is alive and well.
Sep  4 17:07:03 buildnode dnf-distro-sync[3401634]:  Running scriptlet: cvmfs-2.11.0-1.el8.x86_64                              4/7
Sep  4 17:07:04 buildnode dnf-distro-sync[3401634]:  Upgrading        : cvmfs-2.11.0-1.el8.x86_64                              4/7
Sep  4 17:07:58 buildnode systemd[1]: Reloading.
```

**Comment by jblomer** [comment]
Can you give us the full cvmfs syslog of one of these nodes starting from the package update?

**Comment by smuzaffar** [comment]
system log is available in /afs/cern.ch/user/m/muzaffar/public/cvmfs-cmsbuild126.log

**Comment by jblomer** [comment]
Thank you! From what I can see in the logs, it doesn't look like a problem in downloading the whitelist but really in the signature verification itself. Perhaps an issue with loading the public keys or perhaps an issue with symbol resolution, since the 2.11 client links against the libcvmfs_crypto library rather than libcrypto.

Are the affected nodes all running the same OS (which one)?

**Comment by smuzaffar** [comment]
yes all nodes are configured via puppet and run the same OS.
There is another node where cvmfs is still broken. You can find the log file in `~muzaffar/public/cvmfs-cmsbuild117.log`.  CMSSW IB (install in cms-ib.cern.ch repo) `/cvmfs/cms-ib.cern.ch/sw/x86_64/week1/el8_amd64_gcc11/cms/cmssw/CMSSW_13_2_X_2023-09-04-2300` was installed at `Sep  5 03:17` but it is still not visible on this node. Also release installed in cms.cern.ch repo ( /cvmfs/cms.cern.ch/el8_amd64_gcc11/cms/cmssw/CMSSW_13_2_3/) is also not visible.

I can give you admin rights to this node if you need to debug things

**Comment by smuzaffar** [comment]
> Are the affected nodes all running the same OS (which one)?

All nodes are running `AlmaLinux release 8.8 (Sapphire Caracal)`

**Comment by jblomer** [comment]
That's fantastic! Yes, please, let me know the login details. With admin rights, I can then attach a GDB to the cvmfs process.

**Comment by smuzaffar** [comment]
I have sent you the login details vie email

**Comment by jblomer** [comment]
Ok, I see more or less what's happening. The cvmfs-config.cern.ch repository crashed a while ago and it did so in such a way that it doesn't auto-unmount. But instead it keeps the mountpoint alive in a broken state ("transport endpoint not connected"). The log message related to the cvmfs-config.cern.ch crash is this:

```
/var/log/messages-20230820.gz:Aug 15 17:38:12 cmsbuild117.cern.ch cvmfs2[49546]: (cvmfs-config.cern.ch) PANIC: /home/sftnight/jenkins/workspace/CvmfsFullBuildDocker/CVMFS_BUILD_ARCH/docker-x86_64/CVMFS_BUILD_PLATFORM/cc8/build/BUILD/cvmfs-2.10.1/cvmfs/monitor.cc : 522#012watchdog disappeared, aborting
```

These sort of rare crashes should be fixed with cvmfs-2.11: https://github.com/cvmfs/cvmfs/issues/2895

The other fuse processes still think they should get their public keys from the config repository (this is still a bit unclear, see later). Their `CVMFS_KEYS_DIR` environment variable points to the config repository and they have no public keys loaded.

In order to fix nodes in this state in a less intrusive way, you can do

```
umount /cvmfs/cvmfs-config.cern.ch
cvmfs_config reload  # will remount /cvmfs/cvmfs-config.cern.ch along the way
```

There are follow-ups for the cvmfs development:

1. Warn to syslog if no public keys are loaded
2. Investigate why, if the config repository has disappeared, the options manager during reloading does not fall back to the CVMFS_KEYS_DIR location provided in /etc/cvmfs

**Comment by HereThereBeDragons** [comment]
PR that should make more verbosly clear what is wrong is #3387 . It will now crash the mountpoint if neither CVMFS_CONFIG_REPOSITORY or the default config can be loaded. mountpoint now also has extra log message if pubkeys are empty.

side note:
`"/etc/cvmfs/default.conf"` and `dist_default` configs are read the first. after those `CVMFS_CONFIG_REPOSITORY` will be marked as protected.
 no other config (e.g. `/etc/cvmfs/default.local` or `/etc/cvmfs/config.d/*.conf` can overwrite it. 

if a custom `CVMFS_CONFIG_REPOSITORY` is unreachable, now it falls back to the default config (and if that is unreachable: crashes).
(dont think that would have changed here anything)

**Comment by HereThereBeDragons** [comment]
ignore previous comment. real fix is in #3390

**Comment by HereThereBeDragons** [comment]
#3390 should fix it
